sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveTimeWindows.scala

```diff
@@ -257,10 +259,11 @@ object SessionWindowing extends Rule[LogicalPlan] {
           case s: SessionWindow => sessionAttr
         }
 
-        val filterByTimeRange = session.gapDuration match {
-          case Literal(interval: CalendarInterval, CalendarIntervalType) =>
-            interval == null || interval.months + interval.days + interval.microseconds <= 0
-          case _ => true
+        val filterByTimeRange = if (gapDuration.foldable) {
```

Comment:
This is a bit better than pattern matching on a literal since it also supports more complex expressions, in particular `Cast(Literal("<some gap duration>r", StringType), CalendarIntervalType)`. Normally you'd expect the constant folding to take care of this, but this happens during optimization, and this rule is part of analysis; hence the need for manual constant folding.